easypy/typed_struct.py

```diff
@@ -0,0 +1,407 @@
+from copy import deepcopy
+
+from .exceptions import TException
+from .tokens import MANDATORY
+from .collections import ListCollection, PythonOrderedDict
+from .bunch import Bunch, bunchify
+
+
+class InvalidFieldType(TException):
+    template = 'Bad field type {field_type} - {reason}'
+
+
+class NotFields(TException, AttributeError):
+    template = '{field_names} are not fields of {typed_struct.__name__}'
+
+
+class NotAField(TException, AttributeError):
+    template = '{field_name} is not a field of {typed_struct.__name__}'
+
+
+class NotAKey(TException, KeyError):
+    template = '{field_name} is not a key of {typed_struct.__name__}'
+
+
+class MissingField(TException, TypeError):
+    template = '{typed_struct.__name__} is missing mandatory field {field.name}'
+
+
+class FieldTypeMismatch(TException, TypeError):
+    template = '{field.name} is {expected_type}, not {received_type}'
+
+
+class FieldKeyTypeMismatch(TException, TypeError):
+    template = 'keys of {field.name} are {expected_type}, not {received_type}'
+
+
+class Field(object):
+    """
+    Define a single field in a TypedStruct
+
+        type: The type of the field
+              * Use [<type>] for a typechecked list
+              * Use {<key_type>: <type>} for a typechecked dict
+              * Use {str: <type>} for a typechecked Bunch
+        default: Default value for the field
+        preprocess: A function for preprocessing the values assigned to the field:
+                    * If the field can be converted from different types,
+                      should do the conversion and return the converted value.
+                    * If there are validation checks, should perform them and
+                      raise an exception if they fail.
+                    * Use the field.add_validation() and field.add_conversion()
```

Comment:
though... now that I see the api and implementation (with predicate, ex_type, etc.) I think it's quite an overkill and "YAGNI"... But if you insist...

What does @tigrawap  think?